core/src/poh.rs

```diff
@@ -1,93 +1,110 @@
-//! The `Poh` module provhashes an object for generating a Proof of History.
-//! It records Hashes items on behalf of its users.
+//! The `Poh` module provides an object for generating a Proof of History.
 use solana_sdk::hash::{hash, hashv, Hash};
 
 pub struct Poh {
     pub hash: Hash,
     num_hashes: u64,
-    pub tick_height: u64,
+    hashes_per_tick: u64,
+    remaining_hashes: u64,
 }
 
 #[derive(Debug)]
 pub struct PohEntry {
-    pub tick_height: u64,
     pub num_hashes: u64,
     pub hash: Hash,
-    pub mixin: Option<Hash>,
 }
 
 impl Poh {
-    pub fn new(hash: Hash, tick_height: u64) -> Self {
+    pub fn new(hash: Hash, hashes_per_tick: Option<u64>) -> Self {
+        let hashes_per_tick = hashes_per_tick.unwrap_or(std::u64::MAX);
+        assert!(hashes_per_tick > 1);
         Poh {
-            num_hashes: 0,
             hash,
-            tick_height,
+            num_hashes: 0,
+            hashes_per_tick,
+            remaining_hashes: hashes_per_tick,
         }
     }
 
-    pub fn hash(&mut self) {
-        self.hash = hash(&self.hash.as_ref());
-        self.num_hashes += 1;
+    pub fn reset(&mut self, hash: Hash, hashes_per_tick: Option<u64>) {
+        let mut poh = Poh::new(hash, hashes_per_tick);
+        std::mem::swap(&mut poh, self);
     }
 
-    pub fn record(&mut self, mixin: Hash) -> PohEntry {
-        self.hash = hashv(&[&self.hash.as_ref(), &mixin.as_ref()]);
+    pub fn hash(&mut self, max_num_hashes: u64) -> bool {
+        let num_hashes = std::cmp::min(self.remaining_hashes - 1, max_num_hashes);
```

Comment:
I think, the following usecase will call hash more than intended.
```remaining_hashes - 1``` > ```max_num_hashes```
e.g. remaining_hashes = 6, max_num_hashes = 4, it'll hash ```0..4```, i.e. 5 times which is more than ```max_num_hashes```

Not sure if that's a valid scenario for caller of the API, but it can be a potential problem.

Reply:
This should be ok because remaining_hashes will never be larger than max_num_hashes.

But you had me scared for a second and I improved some tests to make me feel better: https://github.com/solana-labs/solana/pull/4351